internal/rpc/otel_interceptor.go

```diff
@@ -0,0 +1,420 @@
+			spec:        conn.Spec(),
+			protocol:    conn.Peer().Protocol,
+			attributes:  attributes,
+			receiveSize: instruments.responseSize,
+			sendSize:    instruments.requestSize,
+		}
+		streamingHandler := &streamingHandlerInterceptor{
+			StreamingHandlerConn: conn,
+			receive: func(msg any, conn connect.StreamingHandlerConn) error {
+				return state.receive(ctx, msg, conn)
+			},
+			send: func(msg any, conn connect.StreamingHandlerConn) error {
+				return state.send(ctx, msg, conn)
+			},
+		}
+		err := next(ctx, streamingHandler)
+		state.attributes = append(
+			state.attributes,
+			statusCodeAttribute(conn.Peer().Protocol, err))
+		if err != nil {
+			span.SetStatus(codes.Error, err.Error())
+		}
+		span.SetAttributes(state.attributes...)
+		instruments.requestsPerRPC.Record(
+			ctx,
+			state.receivedCounter,
+			metric.WithAttributes(state.attributes...))
+		instruments.responsesPerRPC.Record(
+			ctx,
+			state.sentCounter,
+			metric.WithAttributes(state.attributes...))
+		instruments.duration.Record(ctx,
+			time.Since(requestStartTime).Milliseconds(),
+			metric.WithAttributes(state.attributes...))
+		return err
+	}
+}
+
+func propagateOtelHeaders(ctx context.Context, isClient bool, header http.Header) context.Context {
+	if isClient {
+		otel.GetTextMapPropagator().Inject(ctx, propagation.HeaderCarrier(header))
+	} else {
+		ctx = otel.GetTextMapPropagator().Extract(ctx, propagation.HeaderCarrier(header))
+	}
+	return ctx
+}
+
+type instrumentation struct {
+	duration        metric.Int64Histogram
+	requestSize     metric.Int64Histogram
+	responseSize    metric.Int64Histogram
+	requestsPerRPC  metric.Int64Histogram
+	responsesPerRPC metric.Int64Histogram
+}
+
+func createInstruments(meter metric.Meter) instrumentation {
+	duration, err := meter.Int64Histogram("duration", metric.WithUnit("ms"))
+	if err != nil {
+		panic(fmt.Errorf("failed to create duration metric: %w", err))
+	}
+	requestSize, err := meter.Int64Histogram("request.size", metric.WithUnit("By"))
+	if err != nil {
+		panic(fmt.Errorf("failed to create request size metric: %w", err))
+	}
+	responseSize, err := meter.Int64Histogram("response.size", metric.WithUnit("By"))
+	if err != nil {
+		panic(fmt.Errorf("failed to create response size metric: %w", err))
+	}
+	requestsPerRPC, err := meter.Int64Histogram("requests_per_rpc", metric.WithUnit("1"))
+	if err != nil {
+		panic(fmt.Errorf("failed to create requests per rpc metric: %w", err))
+	}
+	responsesPerRPC, err := meter.Int64Histogram("responses_per_rpc", metric.WithUnit("1"))
+	if err != nil {
+		panic(fmt.Errorf("failed to create responses per rpc metric: %w", err))
+	}
+	return instrumentation{
+		duration:        duration,
+		requestSize:     requestSize,
+		responseSize:    responseSize,
+		requestsPerRPC:  requestsPerRPC,
+		responsesPerRPC: responsesPerRPC,
+	}
+}
+
+func statusCodeAttribute(protocol string, err error) attribute.KeyValue {
+	statusCodeKey := fmt.Sprintf("rpc.%s.status_code", protocol)
```

Comment:
Do we need to lowercase protocol, or is it always guaranteed to be consistent case?

Reply:
Using the raw value from the Connection and Spec, so for consistency I'm not lowering here, but in our use it's always just `"grpc"` anyway